src/3rd_party_config.c

```diff
@@ -0,0 +1,235 @@
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/config_parser.h"
+#include "lib/list.h"
+#include "lib/sys.h"
+#include "swupd.h"
+#include <errno.h>
+#include <stdlib.h>
+
+#ifdef THIRDPARTY
+#define MAX_REPO_ALLOWED_FIELDS 2
+
+/** @brief Internal Memory representation of a repo.  */
+struct repo {
+	char *name;
+	char *url;
+};
+
+/** @brief List of repos. */
+struct list *repos_kv;
+
+/**
+ * @brief This function is called by the repo ini parser It is called
+
+ *
+ * @param section A string containing section of the file
+ * @param key A string containing key
+ * @param value A string containing value
+ *
+ * @returns True for every valid line which could be extracted, false if either
+ * section, key or value is NULL
+ */
+bool parse_key_values(char *section, char *key, char *value, void UNUSED_PARAM *data)
+{
+	static int repo_field_count = 0;
+	bool ret = false;
+
+	if (section && key && value) {
+		static struct repo *repo_var;
+		if (repo_field_count == 0) {
+			repo_var = (struct repo *)malloc(sizeof(struct repo));
+			if (repo_var == NULL) {
+				error("Failed to allocate memory");
+				exit(1);
+			}
+
+			repo_var->name = NULL;
+		}
+
+		/* Section name is the name of the repo and needs to only parsed
+		 * the first time
+		 */
+		if (!repo_var->name) {
+			repo_var->name = strdup_or_die(section);
+			repo_field_count++;
+			ret = true;
+		}
+
+		if (!strncmp(key, "url", strlen(key))) {
+			repo_var->url = strdup_or_die(value);
+			repo_field_count++;
+			ret = true;
+		}
+
+		if (repo_field_count == MAX_REPO_ALLOWED_FIELDS) {
+			repos_kv = list_append_data(repos_kv, repo_var);
+			repo_field_count = 0;
+		}
+	}
+	return ret;
+}
+
+static char *get_repo_config_path(void)
+{
+	char *repo_config_file_path;
+	string_or_die(&repo_config_file_path, "%s/%s/%s", globals.state_dir, "3rd_party", "repo.ini");
+	return repo_config_file_path;
+}
+
+static int write_config_repo(struct repo *repo, int mode)
+{
+	int ret = 0;
+	char *repo_string;
+	ssize_t write_line_len;
+	char *repo_config_file_path = get_repo_config_path();
+
+	int fd = open(repo_config_file_path, mode);
+	if (fd < 0) {
+		return -1;
+	}
+
+	string_or_die(&repo_string, "[%s]\nurl=%s\n\n", repo->name, repo->url);
+	write_line_len = strlen(repo_string);
+
+	if (write_line_len != write(fd, repo_string, write_line_len)) {
+		error("error writing to file %d", errno);
+		ret = -1;
+	}
+
+	close(fd);
+	free_string(&repo_string);
+	free_string(&repo_config_file_path);
+	return ret;
+}
+
+static bool repo_config_read(void)
+{
+	bool parse_ret = false;
+	char *repo_config_file_path = get_repo_config_path();
+
+	parse_ret = config_parse(repo_config_file_path, parse_key_values, NULL);
+	free_string(&repo_config_file_path);
+	return parse_ret;
+}
+
+static void repo_config_read_free(void)
+{
+	list_free_list(repos_kv);
```

Comment:
I think you are leaking memory here, you have a list of repo structs, and you are just freeing the list, but the structs are still in memory, and each struct have memory allocated for two strings: name and url. Look at repo_free here -> https://github.com/clearlinux/swupd-client/pull/1192/files#diff-393d66cc4660935afd5f777ffec997a2

Reply:
You are right. Thats what the free_data_function was for. Thanks, I have used what you have adding a todo to move to 3rd-party helpers since I donvi srt know which patch is merging first.

Thanks will change in next push